sw/airborne/firmwares/rotorcraft/guidance/guidance_indi_hybrid.c

```diff
@@ -651,6 +708,98 @@ struct FloatVect3 nav_get_speed_sp_from_line(struct FloatVect2 line_v_enu, struc
 
   return speed_sp_return;
 }
+/**
+ * @brief Go to a waypoint following a controlled diagonal descend towards a rope
+ *
+ * @param target the target waypoint
+ *
+ * @param rope_heading heading in radians of the landing rope calculated from the north
+ * 
+ * @return desired speed FloatVect3
+ */
+struct FloatVect3 nav_get_speed_sp_from_diagonal(struct EnuCoor_i target, float pos_gain, float rope_heading) {
+  // The speed sp that will be returned
+  struct FloatVect3 speed_sp_return = {0, 0, 0};
+  struct NedCoor_f ned_target_wp;
+
+  // Target in NED instead of ENU and altitude relative controller
+  VECT3_ASSIGN(ned_target_wp, POS_FLOAT_OF_BFP(target.y), POS_FLOAT_OF_BFP(target.x), -POS_FLOAT_OF_BFP(target.z));
+  //VECT3_ASSIGN(ned_target_wp, POS_FLOAT_OF_BFP(target.y), POS_FLOAT_OF_BFP(target.x), POS_FLOAT_OF_BFP(-nav_flight_altitude));
+  
+  // Calculate position error
+  struct FloatVect3 pos_error;
+  struct NedCoor_f *pos = stateGetPositionNed_f();
+  VECT3_DIFF(pos_error, ned_target_wp, *pos);
+
+  // Calculate normal vecotor to rope
+  // Rope Heading in degrees
+  struct FloatVect3 n_I;
+  float n_I_N = cosf(rope_heading);
+  float n_I_E = sinf(rope_heading);
+  float n_I_D = 0;
+  VECT3_ASSIGN(n_I,n_I_N,n_I_E,n_I_D);
+
+  // Rotate vector to have a glideslope angle
+  struct FloatVect3 l_I;
+  float c_omega = cosf(RadOfDeg(landing_slope));
+  float s_omega = sinf(RadOfDeg(landing_slope));
+  float l_I_N = -1*n_I.x * c_omega;
+  float l_I_E = -1*n_I.y * c_omega;
+  float l_I_D =  s_omega;
+  VECT3_ASSIGN(l_I,l_I_N , l_I_E , l_I_D);
+
+  // Calculate Vector from projection of drone on GS (Glide Slope) to the rope
+  struct FloatVect3 f_I;
+  float l_I_norm = FLOAT_VECT3_NORM(l_I);
+  float f_ratio = (pos_error.x*l_I.x + pos_error.y*l_I.y+ pos_error.z*l_I.z)/ (l_I_norm * l_I_norm) ;
+  float f_I_N = f_ratio * l_I.x;
+  float f_I_E = f_ratio * l_I.y;
+  float f_I_D = f_ratio * l_I.z;
+  VECT3_ASSIGN(f_I, f_I_N , f_I_E , f_I_D);
+  float f_I_norm = FLOAT_VECT3_NORM(f_I);
+  // Calculate Vector from Drone to Projection on GS
+  struct FloatVect3 d_p;
+  VECT3_DIFF(d_p, pos_error, f_I);
+  float d_p_norm = FLOAT_VECT3_NORM(d_p);
+  // Calculate constant component of velocity SP parallel to GS
+  struct FloatVect3 speed_par;
+  float constant_par = f_I_norm/(f_I_norm+d_p_norm)*pos_gain;//0.01
+  VECT3_SMUL(speed_par, l_I, constant_par);
+
+  // Calculate proportional component of velocity SP perpendicular to GS
+  struct FloatVect3 speed_perp;
+  float test_gain = d_p_norm/(f_I_norm+d_p_norm)*pos_gain;//0.1
```

Comment:
This name is not appropriate. Also, why change the gain?

Reply:
yes the name is not appropriate. Let me change it